### PR TITLE
feat(streaming): external business metrics provider

### PR DESCRIPTION
**Pull request by chengjoey**
#### What this PR does / why we need it:
feat(streaming): external business metrics provider

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/all?id=467790&iterationID=1841&type=TASK)

#### Specified Reviewers:

/assign @sfwn 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English | Support external business metrics provider             |
| 🇨🇳 中文    |      外部业务指标采集的prometheus通用接口        |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: cmd/monitor/collector/bootstrap-agent.yaml [hunk 1]
```diff
@@ -10,7 +10,7 @@ erda.oap.collector.core:
     metrics:
       - receivers:
           #- "erda.oap.collector.receiver.dummy"
-          - "erda.oap.collector.receiver.prometheus-remote-write"
+          - "erda.oap.collector.receiver.prometheus-remote-write@prometheus"
```
Comment (sfwn) [inline cmd/monitor/collector/bootstrap-agent.yaml:13]: `@prometheus` is meaningless. Maybe `@default` or other better name.

File: cmd/monitor/collector/bootstrap-agent.yaml [hunk 2]
```diff
@@ -21,15 +21,33 @@ erda.oap.collector.core:
...
 # ************* receivers *************
 #erda.oap.collector.receiver.dummy:
 #  rate: 3s
 #  metric_sample: '{"name":"kubelet_cadvisor","timeUnixNano":1640936985459000000,"relations":null,"attributes":{"container":"manager","host_ip":"10.118.177.94","id":"/kubepods/burstable/pod164ec226-8106-4904-9bcb-0218a9b2b793/8367a8b0993ebdf8883a0ad8be9c3978b04883e56a156a8de563afa467d49dec","image":"sha256:6cd7bc0e0855164e7ff495c6ec9a37cf8657f8170fe97055ffba2c63343bcedd","instance":"virtual-kubelet-cn-hangzhou-k","name":"8367a8b0993ebdf8883a0ad8be9c3978b04883e56a156a8de563afa467d49dec","namespace":"default","pod":"elasticsearch-operator-776689d978-mjdzq","pod_name":"elasticsearch-operator-776689d978-mjdzq"},"dataPoints":{"container_cpu_usage_seconds_total":13995.161470334,"container_memory_max_usage_bytes":273977344}}'
 
-erda.oap.collector.receiver.prometheus-remote-write:
+erda.oap.collector.receiver.prometheus-remote-write@prometheus:
+
+erda.oap.collector.receiver.prometheus-remote-write@external_metrics:
+  remote_write_url: "${EXTERNAL_METRIC_REMOTE_WRITE_URL:/api/v1/external-remote-write}"
```
Comment (sfwn) [inline cmd/monitor/collector/bootstrap-agent.yaml:50]: `external-remote-write` is only for metric use or any monitor types?
If it's metric only, I think you should make this path clearer.

File: cmd/monitor/collector/bootstrap-agent.yaml [hunk 3]
```diff
@@ -334,6 +352,18 @@ erda.oap.collector.exporter.collector:
     content-encoding: "gzip"
     x-erda-cluster-key: ${DICE_CLUSTER_NAME}
 
+erda.oap.collector.exporter.collector@external_metrics:
+  url: ${COLLECTOR_PUBLIC_URL:http://localhost:7076}/api/v1/collect/external-metrics
+  timeout: ${EXPORTER_COLLECTOR_TIMEOUT:10s}
+  authentication:
+    type: token
+    options:
+      token: ${CLUSTER_ACCESS_KEY:/erda-cluster-credential/CLUSTER_ACCESS_KEY}
```
Comment (sfwn) [inline cmd/monitor/collector/bootstrap-agent.yaml:361]: Is it platform level or cluster level?

File: cmd/monitor/collector/bootstrap.yaml [hunk 4]
```diff
@@ -30,7 +30,7 @@ erda.oap.collector.core:
     metrics:
       - receivers:
           #- "erda.oap.collector.receiver.dummy"
-          - "erda.oap.collector.receiver.prometheus-remote-write"
+          - "erda.oap.collector.receiver.prometheus-remote-write@prometheus"
```
Comment (sfwn) [inline cmd/monitor/collector/bootstrap.yaml:33]: Do we not need to adjust the collector-agent?

File: cmd/monitor/streaming/bootstrap.yaml [hunk 5]
```diff
@@ -371,6 +371,12 @@ storage-retention-strategy@metric:
   load_from_database: true
   ttl_reload_interval: "3m"
 
+storage-retention-strategy@external_metric:
+  # default 180 days
+  default_ttl: "${EXTERNAL_METRIC_TTL:4320h}"
```
Comment (sfwn) [inline cmd/monitor/streaming/bootstrap.yaml:376]: How to keep dop metrics always persisted?

File: cmd/monitor/streaming/bootstrap.yaml [hunk 6]
```diff
@@ -676,6 +692,15 @@ erda.oap.collector.receiver.kafka@spotmetric:
     # TODO. METRIC_CH_GROUP_ID to METRIC_GROUP_ID when all finished
     group: "${METRIC_CH_GROUP_ID:spot-monitor-metrics-ch}"
 
+erda.oap.collector.receiver.kafka@spotexternalmetric:
```
Comment (sfwn) [inline cmd/monitor/streaming/bootstrap.yaml:695]: spot_external_metric

File: cmd/monitor/streaming/bootstrap.yaml [hunk 7]
```diff
@@ -676,6 +692,15 @@ erda.oap.collector.receiver.kafka@spotmetric:
...
+  _enable: ${WRITE_EXTERNAL_METRIC_TO_CLICKHOUSE_ENABLE:true}
+  proto_parser: spotmetric
+  future_discard_date: "${EXTERNAL_METRIC_FUTURE_DISCARD_DATE:24h}"
+  consumer:
+    topics: "${EXTERNAL_METRIC_TOPICS:spot-external-metrics}"
+    # TODO. METRIC_CH_GROUP_ID to METRIC_GROUP_ID when all finished
```
Comment (sfwn) [inline cmd/monitor/streaming/bootstrap.yaml:701]: why todo here

File: cmd/monitor/streaming/conf/clickhouse/external_metrics/ddl_create_default_tables.sql.tpl [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    INDEX idx_metric_source_service_id(tag_values[indexOf(tag_keys, 'source_service_id')]) TYPE bloom_filter GRANULARITY 1,
+    INDEX idx_metric_target_service_id(tag_values[indexOf(tag_keys, 'target_service_id')]) TYPE bloom_filter GRANULARITY 1,
+    INDEX idx_metric_cluster_name(tag_values[indexOf(tag_keys, 'cluster_name')]) TYPE bloom_filter GRANULARITY 1,
+    INDEX idx_container_id(tag_values[indexOf(tag_keys, 'container_id')]) TYPE bloom_filter GRANULARITY 2,
+    INDEX idx_pod_name(tag_values[indexOf(tag_keys, 'pod_name')]) TYPE bloom_filter GRANULARITY 2,
+    INDEX idx_pod_uid(tag_values[indexOf(tag_keys, 'pod_uid')]) TYPE bloom_filter GRANULARITY 2,
+    INDEX idx_metric(tag_values[indexOf(tag_keys, 'metric')]) TYPE bloom_filter GRANULARITY 1,
+    INDEX idx_family_id(tag_values[indexOf(tag_keys, 'family_id')]) TYPE bloom_filter GRANULARITY 3,
+    INDEX idx_terminus_key(tag_values[indexOf(tag_keys, 'terminus_key')]) TYPE bloom_filter GRANULARITY 1
+)
```
Comment (sfwn) [inline cmd/monitor/streaming/conf/clickhouse/external_metrics/ddl_create_default_tables.sql.tpl:24]: How about add `Projection` here?

File: internal/tools/monitor/oap/collector/plugins/exporters/clickhouse/builder/external_metric/external_metrics.go [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+			TagValues:         tagValues,
+		})
+		putnumFieldKeysBuf(numFieldKeys)
+		putnumFieldValuesBuf(numFieldValues)
+		putstrFieldKeysBuf(strFieldKeys)
+		putstrFieldValuesBuf(strFieldValues)
+		puttagKeysBuf(tagKeys)
+		puttagValuesBuf(tagValues)
+		if err1 != nil {
+			_ = batch.Abort()
```
Comment (sfwn) [inline internal/tools/monitor/oap/collector/plugins/exporters/clickhouse/builder/external_metric/external_metrics.go:184]: logs error here

File: internal/tools/monitor/oap/collector/plugins/exporters/clickhouse/builder/external_metric/external_metrics.go [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+		case <-wait: // ignore
+		case <-ctx.Done():
+			return table, nil
+		}
+	}
+	return table, nil
+}
+
+var (
+	numFieldKeysBuf, numFieldValuesBuf, strFieldKeysBuf, strFieldValuesBuf, tagKeysBuf, tagValuesBuf sync.Pool
```
Comment (sfwn) [inline internal/tools/monitor/oap/collector/plugins/exporters/clickhouse/builder/external_metric/external_metrics.go:218]: It's package-level variables. What will happen if we have multiple instances of this provider?

File: internal/tools/monitor/oap/collector/plugins/exporters/clickhouse/builder/external_metric/external_metrics.go [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+	pairsBuf                                                                                         sync.Pool
+)
+
+func getpairsBuf() []odata.Tag {
```
Comment (sfwn) [inline internal/tools/monitor/oap/collector/plugins/exporters/clickhouse/builder/external_metric/external_metrics.go:222]: getPairBuf

**Review by sfwn — CHANGES_REQUESTED** [review]

File: cmd/monitor/collector/bootstrap-agent.yaml (hunk 3 above)
Comment (chengjoey) [reply cmd/monitor/collector/bootstrap-agent.yaml:361]: cluster level

File: internal/tools/monitor/oap/collector/plugins/exporters/clickhouse/builder/external_metric/external_metrics.go (hunk 10 above)
Comment (chengjoey) [reply internal/tools/monitor/oap/collector/plugins/exporters/clickhouse/builder/external_metric/external_metrics.go:218]: metrics will be confused🤪

File: internal/tools/monitor/oap/collector/plugins/exporters/clickhouse/builder/external_metric/external_metrics.go (hunk 11 above)
Comment (chengjoey) [reply internal/tools/monitor/oap/collector/plugins/exporters/clickhouse/builder/external_metric/external_metrics.go:222]: adjusted

File: internal/tools/monitor/oap/collector/plugins/exporters/clickhouse/builder/external_metric/external_metrics.go (hunk 9 above)
Comment (chengjoey) [reply internal/tools/monitor/oap/collector/plugins/exporters/clickhouse/builder/external_metric/external_metrics.go:184]: added

File: cmd/monitor/collector/bootstrap-agent.yaml (hunk 1 above)
Comment (chengjoey) [reply cmd/monitor/collector/bootstrap-agent.yaml]: adjusted, use `default`

File: cmd/monitor/collector/bootstrap-agent.yaml (hunk 2 above)
Comment (chengjoey) [reply cmd/monitor/collector/bootstrap-agent.yaml:50]: remote_write_url it's used for prometheus, so is obvious, that is to use for metric

File: cmd/monitor/collector/bootstrap.yaml (hunk 4 above)
Comment (chengjoey) [reply cmd/monitor/collector/bootstrap.yaml:33]: not need

File: cmd/monitor/streaming/bootstrap.yaml (hunk 5 above)
Comment (chengjoey) [reply cmd/monitor/streaming/bootstrap.yaml]: removed this  useless provider

File: cmd/monitor/streaming/bootstrap.yaml (hunk 6 above)
Comment (chengjoey) [reply cmd/monitor/streaming/bootstrap.yaml]: adjusted

File: cmd/monitor/streaming/bootstrap.yaml (hunk 7 above)
Comment (chengjoey) [reply cmd/monitor/streaming/bootstrap.yaml]: removed

File: cmd/monitor/streaming/conf/clickhouse/external_metrics/ddl_create_default_tables.sql.tpl (hunk 8 above)
Comment (chengjoey) [reply cmd/monitor/streaming/conf/clickhouse/external_metrics/ddl_create_default_tables.sql.tpl:25]: add idx_project_id

File: cmd/monitor/collector/bootstrap-agent.yaml (hunk 2 above)
Comment (chengjoey) [reply cmd/monitor/collector/bootstrap-agent.yaml]: adjusted to `/api/v1/external-prometheus-remote-write`

**Review by sfwn — APPROVED** [review]

**Pull request merged**